### ISSUE TITLE
awful new slurm GresUsed format breaks pyslurm

### ISSUE THREAD

**Issue by wresch** [PySlurm/pyslurm#104]
#### Details

* Slurm Version: 17.02.9
* Python Version: 2.6
* Cython Version: 0.26
* PySlurm Branch: 17.02
* Linux Distribution: Centos6

#### Issue

Starting with slurm 17 (https://github.com/SchedMD/slurm/blob/85df676e3945d4464ecfbc501e758670752fece9/NEWS#L1295), GresUsed can now take the format
```
GresUsed=gpu:p100:2(IDX:1,3),lscratch:0
```
which is just awful b/c it overloads the colon and comma inside a nested list.  And unfortunately that awfulness breaks  `slurm.listOrNone` for gres strings

```
cdef inline listOrNone(char* value, sep_char):
    if value is NULL:
        return []

    if not sep_char:
        return value.decode("UTF-8", "replace")

    if sep_char == b'':
        return value.decode("UTF-8", "replace")

return value.decode("UTF_8", "replace").split(sep_char)
```

is called with `slurm.listOrNone(gresstr, ',')`, which now fails for for such gres lists. Well - it doesn't fail - it returns ` ['gpu:p100:2(IDX:1', ',3)', ,'lscratch:0']`

**Comment by giovtorres** [comment]
Yeah, that's not very friendly to parse.  Perhaps we could return it as a string instead of a list.  Remind me, without those file IDs, the pyslurm output is just a list with a single string?

**Comment by wresch** [comment]
pyslurm returns `slurm.listOrNone(gresstr, ',')`, so a list with one entry per gres type if it did not have that extra stuff. So logically for the gres string 
```
GresUsed=gpu:p100:2(IDX:1,3),lscratch:0
```
i think it should return
```
['gpu:p100:2(IDX:1,3)', 'lscratch:0']
```

or i suppose as you suggest it could just return the string unmodified. not sure how much stuff downstream depends on interpreting that string.

**Comment by wresch** [comment]
i can create a pull request that will ignore commas inside parens, if you'd like

**Comment by giovtorres** [comment]
Sure, that would be great @wresch. Thanks!

**Comment by giovtorres** [comment]
@wresch did you manage to put together something for this issue?

**Comment by wresch** [comment]
This weekend

**Comment by chrissamuel** [comment]
HI all,

Just wondered whether anyone was working on this?  We'd like to be able to add more support for GPUs in our job monitor application (linked above) but we're dependent upon this.  Unfortunately I don't speak Python so I'm not able to help out. :-(

All the best,
Chris

**Comment by giovtorres** [comment]
@chrissamuel Would you mind sending me some samples out the output for `GresUsed`.  I don't have access to a cluster at the moment.  I can write a parser based on the various outputs.  Thanks.

**Comment by chrissamuel** [comment]
Hi Giovanni,

On 02/07/18 12:04, Giovanni Torres wrote:

> @chrissamuel <https://github.com/chrissamuel> Would you mind sending me 
> some samples out the output for |GresUsed|. I don't have access to a 
> cluster at the moment. I can write a parser based on the various 
> outputs. Thanks.

Currently we've only got 2 GPU jobs running (usage is only just starting 
to ramp up again after the weekend).  They both have:

# scontrol show job -d | fgrep -i GRES_IDX | egrep -v 'GRES_IDX=$'
      Nodes=john85 CPU_IDs=4-11 Mem=40960 GRES_IDX=gpu(IDX:0)
      Nodes=john1 CPU_IDs=15 Mem=8192 GRES_IDX=gpu(IDX:0)

So it's the GRES_IDX we need to use.

Submitting a job that asks for 2 GPUs shows:

      Nodes=john101 CPU_IDs=16 Mem=100 GRES_IDX=gpu(IDX:0-1)

We only have 2 GPUs per node so I can't show what it looks like if
I have a non-contiguous range of GPUs allocated I'm afraid.

Thanks so much for this!

All the best,
Chris
-- 
  Chris Samuel  :  http://www.csamuel.org/  :  Melbourne, VIC

**Comment by giovtorres** [comment]
I think this issue is different from what you're seeking.  You're looking for the `GRES_IDX` value.  If that's so, could you open another issue for this missing value.

**Comment by chrissamuel** [comment]
On Tuesday, 3 July 2018 10:27:25 AM AEST Giovanni Torres wrote:

> >>> import pyslurm
> >>> pyslurm.version()

'17.11.0'

-- 
 Chris Samuel  :  http://www.csamuel.org/  :  Melbourne, VIC

**Comment by chrissamuel** [comment]
Ah, I see no `GresUsed` in the output of `scontrol show job -d` so I was presuming that this was what you were mapping `GRES_IDX` to.  I've just checked the source and `GresUsed` is the node attribute whilst `GRES_IDX` is the job attribute, hence my confusion.

I have to run now, but I'll open a new issue for that later today.

**Comment by chrissamuel** [comment]
On Monday, 2 July 2018 12:04:55 PM AEST Giovanni Torres wrote:

> @chrissamuel Would you mind sending me some samples out the output for
> `GresUsed`.  I don't have access to a cluster at the moment.  I can write a
> parser based on the various outputs.  Thanks.

Here is what you were after:

# scontrol show node -d | fgrep GresUsed | sort | uniq
   GresUsed=gpu:0,hbm:0
   GresUsed=gpu:p100:0(IDX:N/A),hbm:0
   GresUsed=gpu:p100:1(IDX:0),hbm:0
   GresUsed=gpu:p100:1(IDX:1),hbm:0
   GresUsed=gpu:p100:2(IDX:0-1),hbm:0

The first line will be from our 4 KNL nodes that have no GPUs (the Skylake 
nodes have dual P100's).

-- 
 Chris Samuel  :  http://www.csamuel.org/  :  Melbourne, VIC

**Comment by giovtorres** [comment]
@chrissamuel Thanks for these examples.  I've used them as test cases for the regex.

**Comment by giovtorres** [comment]
Fixed in #117 .   Please reopen if it is still an issue.

/cc @wresch

**Comment by chrissamuel** [comment]
Thanks!  Sorry I didn't see this update before now, I've no idea why!
